Connection pool is shut down

## Description

Getting `Connection pool is shut down` after some time. Seems to be a Jersey issue.
## How to reproduce

We have a daemon process using the client like that:

```
    private void request() {
         try (DockerClient dockerClient = getDockerClient()) {
         } catch (Exception e) {
             //Log exception
         }
    }

    private static DockerClient getDockerClient() {
        final long TIMEOUT = 5 * 60 * 1000;
        return DefaultDockerClient.builder()
                .uri(URI.create("http://url/"))
                .readTimeoutMillis(TIMEOUT)
                .build();
    }

```
## What do you expect

Client should continue to work.
## What happened instead

Client connection pool is shut  down.
## Software:
- docker version: 1.11
- docker-client version: 3.6.8
## Full backtrace

``` text
com.spotify.docker.client.DockerException: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.IllegalStateException: Connection pool shut down
        at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1488)
        at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1441)
        at com.spotify.docker.client.DefaultDockerClient.removeContainer(DefaultDockerClient.java:661)
        at <some our classes>
Caused by: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.IllegalStateException: Connection pool shut down
        at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299)
        at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:286)
        at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
        at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1439)
        ... 11 more
Caused by: javax.ws.rs.ProcessingException: java.lang.IllegalStateException: Connection pool shut down
        at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:481)
        at org.glassfish.jersey.apache.connector.ApacheConnector$1.run(ApacheConnector.java:491)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at jersey.repackaged.com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:299)
        at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:112)
        at jersey.repackaged.com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:50)
        at jersey.repackaged.com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:37)
        at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:487)
        at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:177)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:340)
        at org.glassfish.jersey.client.ClientRuntime$3.run(ClientRuntime.java:209)
        ... 5 more
Caused by: java.lang.IllegalStateException: Connection pool shut down
        at org.apache.http.util.Asserts.check(Asserts.java:34)
        at org.apache.http.pool.AbstractConnPool.lease(AbstractConnPool.java:184)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.requestConnection(PoolingHttpClientConnectionManager.java:251)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:175)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
        at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:435)
        ... 21 more
```

--- a/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
+++ b/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
@@ -10,9 +10,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *      http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -23,40 +23,28 @@
 
 package com.spotify.docker.client;
 
-import static com.google.common.base.MoreObjects.firstNonNull;
-import static com.google.common.base.Optional.fromNullable;
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Preconditions.checkNotNull;
-import static com.google.common.base.Strings.isNullOrEmpty;
-import static com.google.common.collect.Maps.newHashMap;
-import static com.spotify.docker.client.ObjectMapperProvider.objectMapper;
-import static com.spotify.docker.client.VersionCompare.compareVersion;
-import static java.nio.charset.StandardCharsets.UTF_8;
-import static java.util.Collections.singletonMap;
-import static java.util.concurrent.TimeUnit.SECONDS;
-import static javax.ws.rs.HttpMethod.DELETE;
-import static javax.ws.rs.HttpMethod.GET;
-import static javax.ws.rs.HttpMethod.POST;
-import static javax.ws.rs.HttpMethod.PUT;
-import static javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE;
-import static javax.ws.rs.core.MediaType.APPLICATION_OCTET_STREAM;
-import static javax.ws.rs.core.MediaType.APPLICATION_OCTET_STREAM_TYPE;
-import static javax.ws.rs.core.MediaType.TEXT_PLAIN_TYPE;
-
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.node.JsonNodeFactory;
 import com.google.common.annotations.VisibleForTesting;
+import static com.google.common.base.MoreObjects.firstNonNull;
 import com.google.common.base.Optional;
+import static com.google.common.base.Optional.fromNullable;
 import com.google.common.base.Preconditions;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 import com.google.common.base.Strings;
+import static com.google.common.base.Strings.isNullOrEmpty;
 import com.google.common.base.Supplier;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+import static com.google.common.collect.Maps.newHashMap;
 import com.google.common.io.CharStreams;
 import com.google.common.net.HostAndPort;
+import static com.spotify.docker.client.ObjectMapperProvider.objectMapper;
+import static com.spotify.docker.client.VersionCompare.compareVersion;
 import com.spotify.docker.client.exceptions.BadParamException;
 import com.spotify.docker.client.exceptions.ConflictException;
 import com.spotify.docker.client.exceptions.ContainerNotFoundException;
@@ -125,12 +113,13 @@ import java.io.UnsupportedEncodingException;
 import java.net.SocketTimeoutException;
 import java.net.URI;
 import java.net.URLEncoder;
+import static java.nio.charset.StandardCharsets.UTF_8;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.AbstractMap;
 import java.util.ArrayList;
 import java.util.Collections;
+import static java.util.Collections.singletonMap;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -139,8 +128,13 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
+import static java.util.concurrent.TimeUnit.SECONDS;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import static javax.ws.rs.HttpMethod.DELETE;
+import static javax.ws.rs.HttpMethod.GET;
+import static javax.ws.rs.HttpMethod.POST;
+import static javax.ws.rs.HttpMethod.PUT;
 import javax.ws.rs.ProcessingException;
 import javax.ws.rs.WebApplicationException;
 import javax.ws.rs.client.Client;
@@ -150,6 +144,10 @@ import javax.ws.rs.client.Invocation;
 import javax.ws.rs.client.ResponseProcessingException;
 import javax.ws.rs.client.WebTarget;
 import javax.ws.rs.core.GenericType;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE;
+import static javax.ws.rs.core.MediaType.APPLICATION_OCTET_STREAM;
+import static javax.ws.rs.core.MediaType.APPLICATION_OCTET_STREAM_TYPE;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN_TYPE;
 import javax.ws.rs.core.MultivaluedHashMap;
 import javax.ws.rs.core.MultivaluedMap;
 import javax.ws.rs.core.Response;
@@ -218,7 +216,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     }
 
   }
-  
+
   /**
    * Hack: this {@link ProgressHandler} is meant to capture the image names
    * of an image being loaded. Weirdly enough, Docker returns the name of a newly
@@ -253,7 +251,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
         if (streamMatcher.matches()) {
           imageNames.add(streamMatcher.group("image"));
         }
-        
+
       }
     }
 
@@ -271,11 +269,12 @@ public class DefaultDockerClient implements DockerClient, Closeable {
   private static final long DEFAULT_READ_TIMEOUT_MILLIS = SECONDS.toMillis(30);
   private static final int DEFAULT_CONNECTION_POOL_SIZE = 100;
 
-  private static final ClientConfig DEFAULT_CONFIG = new ClientConfig(
+  private final ClientConfig defaultConfig = new ClientConfig(
       ObjectMapperProvider.class,
       JacksonFeature.class,
       LogsResponseReader.class,
-      ProgressResponseReader.class);
+      ProgressResponseReader.class
+  );
 
   private static final Pattern CONTAINER_NAME_PATTERN =
           Pattern.compile("^[a-zA-Z0-9][a-zA-Z0-9_.-]+$");
@@ -323,7 +322,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
   private static final GenericType<List<Task>> TASK_LIST = new GenericType<List<Task>>() { };
 
   private static final GenericType<List<Node>> NODE_LIST = new GenericType<List<Node>>() { };
-  
+
   private static final GenericType<List<Secret>> SECRET_LIST = new GenericType<List<Secret>>() { };
 
   private final Client client;
@@ -405,7 +404,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
         .setSocketTimeout((int) builder.readTimeoutMillis)
         .build();
 
-    final ClientConfig config = DEFAULT_CONFIG
+    final ClientConfig config = defaultConfig
         .connectorProvider(new ApacheConnectorProvider())
         .property(ApacheClientProperties.CONNECTION_MANAGER, cm)
         .property(ApacheClientProperties.REQUEST_CONFIG, requestConfig);
@@ -1101,7 +1100,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
       throws DockerException, InterruptedException {
     create(image, imagePayload, handler);
   }
-  
+
   @Override
   public Set<String> load(final InputStream imagePayload)
       throws DockerException, InterruptedException {
@@ -1115,10 +1114,10 @@ public class DefaultDockerClient implements DockerClient, Closeable {
             .path("images")
             .path("load")
             .queryParam("quiet", "false");
-    
+
     final LoadProgressHandler loadProgressHandler = new LoadProgressHandler(handler);
     final Entity<InputStream> entity = Entity.entity(imagePayload, APPLICATION_OCTET_STREAM);
-    
+
     try (final ProgressStream load =
             request(POST, ProgressStream.class, resource,
                     resource.request(APPLICATION_JSON_TYPE), entity)) {
@@ -1831,7 +1830,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     WebTarget resource = resource().path("nodes");
     return request(GET, NODE_LIST, resource, resource.request(APPLICATION_JSON_TYPE));
   }
-  
+
   @Override
   public void execResizeTty(final String execId,
                             final Integer height,
@@ -1935,7 +1934,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     resource = addParameters(resource, params);
     return request(GET, NETWORK_LIST, resource, resource.request(APPLICATION_JSON_TYPE));
   }
-  
+
   @Override
   public Network inspectNetwork(String networkId) throws DockerException, InterruptedException {
     final WebTarget resource = resource().path("networks").path(networkId);

--- a/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
+++ b/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
@@ -10,9 +10,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- *
+ * 
  *      http://www.apache.org/licenses/LICENSE-2.0
- *
+ * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

--- a/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
+++ b/src/main/java/com/spotify/docker/client/DefaultDockerClient.java
@@ -10,9 +10,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *      http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -222,7 +222,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     }
 
   }
-  
+
   /**
    * Hack: this {@link ProgressHandler} is meant to capture the image names
    * of an image being loaded. Weirdly enough, Docker returns the name of a newly
@@ -257,7 +257,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
         if (streamMatcher.matches()) {
           imageNames.add(streamMatcher.group("image"));
         }
-        
+
       }
     }
 
@@ -275,7 +275,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
   private static final long DEFAULT_READ_TIMEOUT_MILLIS = SECONDS.toMillis(30);
   private static final int DEFAULT_CONNECTION_POOL_SIZE = 100;
 
-  private static final ClientConfig DEFAULT_CONFIG = new ClientConfig(
+  private final ClientConfig defaultConfig = new ClientConfig(
       ObjectMapperProvider.class,
       JacksonFeature.class,
       LogsResponseReader.class,
@@ -327,7 +327,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
   private static final GenericType<List<Task>> TASK_LIST = new GenericType<List<Task>>() { };
 
   private static final GenericType<List<Node>> NODE_LIST = new GenericType<List<Node>>() { };
-  
+
   private static final GenericType<List<Secret>> SECRET_LIST = new GenericType<List<Secret>>() { };
 
   private final Client client;
@@ -409,7 +409,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
         .setSocketTimeout((int) builder.readTimeoutMillis)
         .build();
 
-    final ClientConfig config = DEFAULT_CONFIG
+    final ClientConfig config = defaultConfig
         .connectorProvider(new ApacheConnectorProvider())
         .property(ApacheClientProperties.CONNECTION_MANAGER, cm)
         .property(ApacheClientProperties.REQUEST_CONFIG, requestConfig);
@@ -1103,10 +1103,10 @@ public class DefaultDockerClient implements DockerClient, Closeable {
             .path("images")
             .path("load")
             .queryParam("quiet", "false");
-    
+
     final LoadProgressHandler loadProgressHandler = new LoadProgressHandler(handler);
     final Entity<InputStream> entity = Entity.entity(imagePayload, APPLICATION_OCTET_STREAM);
-    
+
     try (final ProgressStream load =
             request(POST, ProgressStream.class, resource,
                     resource.request(APPLICATION_JSON_TYPE), entity)) {
@@ -1975,7 +1975,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     WebTarget resource = resource().path("nodes");
     return request(GET, NODE_LIST, resource, resource.request(APPLICATION_JSON_TYPE));
   }
-  
+
   @Override
   public void execResizeTty(final String execId,
                             final Integer height,
@@ -2079,7 +2079,7 @@ public class DefaultDockerClient implements DockerClient, Closeable {
     resource = addParameters(resource, params);
     return request(GET, NETWORK_LIST, resource, resource.request(APPLICATION_JSON_TYPE));
   }
-  
+
   @Override
   public Network inspectNetwork(String networkId) throws DockerException, InterruptedException {
     final WebTarget resource = resource().path("networks").path(networkId);

--- a/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
+++ b/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
@@ -3,14 +3,13 @@
  * docker-client
  * --
  * Copyright (C) 2016 Spotify AB
- * Copyright (C) 2016 Thoughtworks, Inc
  * --
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- *
+ * 
  *      http://www.apache.org/licenses/LICENSE-2.0
- *
+ * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -27,6 +26,9 @@ import com.spotify.docker.client.exceptions.DockerCertificateException;
 import com.spotify.docker.client.exceptions.DockerException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ThreadLocalRandom;
 import org.hamcrest.MatcherAssert;

--- a/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
+++ b/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
@@ -21,14 +21,12 @@
 
 package com.spotify.docker.client;
 
+import static java.lang.Long.toHexString;
+
 import com.spotify.docker.client.exceptions.DockerCertificateException;
 import com.spotify.docker.client.exceptions.DockerException;
-import static java.lang.Long.toHexString;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.Callable;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ThreadLocalRandom;
 import org.hamcrest.MatcherAssert;
@@ -42,63 +40,64 @@ import org.slf4j.LoggerFactory;
 
 /**
  * Checks thread safety of DefaultDockerClient implementation.
+ *
  * @author Umberto Nicoletti (umberto.nicoletti@gmail.com)
  */
 public class ConnectionPoolTest {
 
-    private static final String BUSYBOX = "busybox";
-    private static final String BUSYBOX_LATEST = BUSYBOX + ":latest";
-    private static final String BUSYBOX_BUILDROOT_2013_08_1 = BUSYBOX + ":buildroot-2013.08.1";
-    private static final String MEMCACHED = "rohan/memcached-mini";
-    private static final String CIRROS_PRIVATE = "dxia/cirros-private";
+  private static final String BUSYBOX = "busybox";
+  private static final String BUSYBOX_LATEST = BUSYBOX + ":latest";
+  private static final String BUSYBOX_BUILDROOT_2013_08_1 = BUSYBOX + ":buildroot-2013.08.1";
+  private static final String MEMCACHED = "rohan/memcached-mini";
+  private static final String CIRROS_PRIVATE = "dxia/cirros-private";
 
-    private static final Logger log = LoggerFactory.getLogger(ConnectionPoolTest.class);
+  private static final Logger log = LoggerFactory.getLogger(ConnectionPoolTest.class);
 
-    @Rule
-    public final ExpectedException exception = ExpectedException.none();
+  @Rule
+  public final ExpectedException exception = ExpectedException.none();
 
-    @Rule
-    public final TestName testName = new TestName();
+  @Rule
+  public final TestName testName = new TestName();
 
-    private final String nameTag = toHexString(ThreadLocalRandom.current().nextLong());
+  private final String nameTag = toHexString(ThreadLocalRandom.current().nextLong());
 
-    /**
-     * Checks that running a parallel operation does not break DefaultDockerClient.
-     * Fixes issue #446.
-     *
-     * @throws Exception on error.
-     */
-    @Test
-    public void testParallelOperation() throws Exception {
-        final ExecutorService executor = Executors.newFixedThreadPool(5);
-        List<Future<Exception>> tasks = new ArrayList<>(20);
-        for(int i=0; i<20; i++) {
-            tasks.add(executor.submit(
-                    new Callable<Exception>() {
-                        @Override
-                        public Exception call() throws Exception {
-                            try (DockerClient docker = DefaultDockerClient.fromEnv().build()) {
-                                docker.pull(ConnectionPoolTest.BUSYBOX_LATEST);
-                                docker.pull(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1);
-                                final String name = "test-repo/tag-force:sometag";
-                                docker.tag(ConnectionPoolTest.BUSYBOX_LATEST, name);
-                                docker.tag(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1, name, true);
-                            } catch (InterruptedException | DockerException | DockerCertificateException e) {
-                                ConnectionPoolTest.log.error(
-                                    "Error running task: {}", e.getMessage(), e
-                                );
-                                return e;
-                            }
-                            return null;
-                        }
-                    }
-                )
-            );
-        }
-        executor.shutdown();
-        while(!executor.isTerminated());
-        for (final Future<Exception> task: tasks) {
-            MatcherAssert.assertThat(task.get(), Matchers.nullValue());
-        }
+  /**
+   * Checks that running a parallel operation does not break DefaultDockerClient.
+   * Fixes issue #446.
+   *
+   * @throws Exception on error.
+   */
+  @Test
+  public void testParallelOperation() throws Exception {
+    final ExecutorService executor = Executors.newFixedThreadPool(5);
+    List<Future<Exception>> tasks = new ArrayList<>(20);
+    for (int i = 0; i < 20; i++) {
+      tasks.add(executor.submit(
+              new Callable<Exception>() {
+                @Override
+                public Exception call() throws Exception {
+                  try (DockerClient docker = DefaultDockerClient.fromEnv().build()) {
+                    docker.pull(ConnectionPoolTest.BUSYBOX_LATEST);
+                    docker.pull(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1);
+                    final String name = "test-repo/tag-force:sometag";
+                    docker.tag(ConnectionPoolTest.BUSYBOX_LATEST, name);
+                    docker.tag(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1, name, true);
+                  } catch (InterruptedException | DockerException | DockerCertificateException e) {
+                    ConnectionPoolTest.log.error(
+                            "Error running task: {}", e.getMessage(), e
+                    );
+                    return e;
+                  }
+                  return null;
+                }
+              }
+              )
+      );
     }
+    executor.shutdown();
+    while (!executor.isTerminated()) {};
+    for (final Future<Exception> task : tasks) {
+      MatcherAssert.assertThat(task.get(), Matchers.nullValue());
+    }
+  }
 }

--- a/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
+++ b/src/test/java/com/spotify/docker/client/ConnectionPoolTest.java
@@ -1,0 +1,104 @@
+/*-
+ * -\-\-
+ * docker-client
+ * --
+ * Copyright (C) 2016 Spotify AB
+ * Copyright (C) 2016 Thoughtworks, Inc
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.docker.client;
+
+import com.spotify.docker.client.exceptions.DockerCertificateException;
+import com.spotify.docker.client.exceptions.DockerException;
+import static java.lang.Long.toHexString;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TestName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checks thread safety of DefaultDockerClient implementation.
+ * @author Umberto Nicoletti (umberto.nicoletti@gmail.com)
+ */
+public class ConnectionPoolTest {
+
+    private static final String BUSYBOX = "busybox";
+    private static final String BUSYBOX_LATEST = BUSYBOX + ":latest";
+    private static final String BUSYBOX_BUILDROOT_2013_08_1 = BUSYBOX + ":buildroot-2013.08.1";
+    private static final String MEMCACHED = "rohan/memcached-mini";
+    private static final String CIRROS_PRIVATE = "dxia/cirros-private";
+
+    private static final Logger log = LoggerFactory.getLogger(ConnectionPoolTest.class);
+
+    @Rule
+    public final ExpectedException exception = ExpectedException.none();
+
+    @Rule
+    public final TestName testName = new TestName();
+
+    private final String nameTag = toHexString(ThreadLocalRandom.current().nextLong());
+
+    /**
+     * Checks that running a parallel operation does not break DefaultDockerClient.
+     * Fixes issue #446.
+     *
+     * @throws Exception on error.
+     */
+    @Test
+    public void testParallelOperation() throws Exception {
+        final ExecutorService executor = Executors.newFixedThreadPool(5);
+        List<Future<Exception>> tasks = new ArrayList<>(20);
+        for(int i=0; i<20; i++) {
+            tasks.add(executor.submit(
+                    new Callable<Exception>() {
+                        @Override
+                        public Exception call() throws Exception {
+                            try (DockerClient docker = DefaultDockerClient.fromEnv().build()) {
+                                docker.pull(ConnectionPoolTest.BUSYBOX_LATEST);
+                                docker.pull(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1);
+                                final String name = "test-repo/tag-force:sometag";
+                                docker.tag(ConnectionPoolTest.BUSYBOX_LATEST, name);
+                                docker.tag(ConnectionPoolTest.BUSYBOX_BUILDROOT_2013_08_1, name, true);
+                            } catch (InterruptedException | DockerException | DockerCertificateException e) {
+                                ConnectionPoolTest.log.error(
+                                    "Error running task: {}", e.getMessage(), e
+                                );
+                                return e;
+                            }
+                            return null;
+                        }
+                    }
+                )
+            );
+        }
+        executor.shutdown();
+        while(!executor.isTerminated());
+        for (final Future<Exception> task: tasks) {
+            MatcherAssert.assertThat(task.get(), Matchers.nullValue());
+        }
+    }
+}

